core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -231,7 +231,9 @@ public void commit() {
             Snapshot newSnapshot = apply();
             newSnapshotId.set(newSnapshot.snapshotId());
             TableMetadata updated = base.replaceCurrentSnapshot(newSnapshot);
-            taskOps.commit(base, updated);
+            // if the table UUID is missing, add it here. the UUID will be re-created each time this operation retries
+            // to ensure that if a concurrent operation assigns the UUID, this operation will not fail.
+            taskOps.commit(base, updated.withUUID());
```

Comment:
Just to make sure I understand: table UUID isn't supposed to change once assigned and cannot be set back to null, right?

Reply:
Correct. This identifies a table so you can't start an operation, drop and recreate the table, and then complete the operation. We had it happen once.